src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp

```diff
@@ -11,87 +14,376 @@
+    setPyException(cppExc, pyModule, CppExceptionsAndPyClassNames::kClassNames[0]);
+  }
+}
+
+/// Mapping of Python class name to Python class object.
+using PyObjectByName = std::map<const std::string_view, const py::object>;
+
+/**
+ * Register a new Python exception.
+ *
+ * General case of a simple exception (only takes a single string
+ * constructor argument) with OpenAssetIOException as its base class.
+ *
+ * @tparam Exception Corresponding C++ exception type.
+ * @param pyClassName Python exception class name to register.
+ * @param mod Python module to hold new Python exception class.
+ * @param registeredExceptions Previously registered exceptions, to use
+ * when looking up Python base class to inherit from, and to add this
+ * exception to once registered.
+ */
+template <class Exception>
+void registerPyExceptionClass(const std::string_view pyClassName, const py::module_ &mod,
+                              PyObjectByName &registeredExceptions) {
+  const py::object pyBaseClass = registeredExceptions.at("OpenAssetIOException");
+  // Register the exception via pybind11.
+  py::exception<void /* unused */> pyExc{mod, pyClassName.data(), pyBaseClass};
+  registeredExceptions.insert({pyClassName, std::move(pyExc)});
+}
+/**
+ * Register a new OpenAssetIOException Python exception.
+ *
+ * Specialisation for OpenAssetIOException root base class, which
+ * inherits from built-in Python RuntimeError exception.
+ *
+ * @param pyClassName Python exception class name to register.
+ * @param mod Python module to hold new Python exception class.
+ * @param registeredExceptions Previously registered exceptions, to add
+ * this exception to once registered.
+ */
+template <>
+void registerPyExceptionClass<OpenAssetIOException>(const std::string_view pyClassName,
+                                                    const py::module_ &mod,
+                                                    PyObjectByName &registeredExceptions) {
+  py::exception<void /* unused */> pyExc{mod, pyClassName.data(), PyExc_RuntimeError};
+  registeredExceptions.insert({pyClassName, std::move(pyExc)});
+}
+/**
+ * Register a new ConfigurationException Python exception.
+ *
+ * Specialisation for ConfigurationException, which inherits from
+ * InputValidationException.
+ *
+ * @param pyClassName Python exception class name to register.
+ * @param mod Python module to hold new Python exception class.
+ * @param registeredExceptions Previously registered exceptions, to use
+ * when looking up Python base class to inherit from, and to add this
+ * exception to once registered.
+ */
+template <>
+void registerPyExceptionClass<ConfigurationException>(const std::string_view pyClassName,
+                                                      const py::module_ &mod,
+                                                      PyObjectByName &registeredExceptions) {
+  const py::object pyBaseClass = registeredExceptions.at("InputValidationException");
```

Comment:
[question][naive] Thinking out loud, from a very uninformed starting point, could you use `kCppExceptionsAndPyClassNames` to auto-derive `pyBaseClass` from the introspected base of the template so you dont need a specialisations? Might not be worth it.

Reply:
Heh, i noted in my review how I thought resisting the urge to do this was a good call :P.
Either works, but don't let the genericism override the clarity.

Reply:
Had some fun in f4993c24 - can drop if you think it's excessive (it only helps remove a single case). It does have the neat side-effect of ensuring that the Python class hierarchy mirrors C++.

Reply:
> It does have the neat side-effect of ensuring that the Python class hierarchy mirrors C++.

For me thats a useful benefit, as it makes future maintenance easier and less prone to human error. @elliotcmorris what are your thoughts?

Reply:
> It does have the neat side-effect of ensuring that the Python class hierarchy mirrors C++.

Suppose that's fair. It's not really that much more complicated either.